lib/httpigeon/circuit_breaker/fuse.rb

```diff
@@ -0,0 +1,205 @@
+require_relative 'errors'
+require_relative 'fuse_config'
+require_relative 'memory_store'
+require_relative '../middleware/circuit_breaker'
+
+module HTTPigeon
+  module CircuitBreaker
+    class Fuse
+      STATE_OPEN = 'open'.freeze
+      STATE_HALF_OPEN = 'half_open'.freeze
+      STATE_CLOSED = 'closed'.freeze
+
+      def self.from_options(options)
+        new(FuseConfig.new(options))
+      end
+
+      attr_reader :service_id, :config, :storage
+
+      def initialize(config)
+        @config = config
+        @service_id = config.service_id.to_s
+        @storage = CircuitBreaker::MemoryStore.new(config.sample_window)
+        @open_storage_key = "circuit:#{service_id}:#{STATE_OPEN}"
+        @half_open_storage_key = "circuit:#{service_id}:#{STATE_HALF_OPEN}"
+        @state_change_syncer = Mutex.new
+      end
+
+      def execute(request_id: nil)
+        @request_id = request_id
+
+        if open?
+          record_tripped!
+
+          config.on_open_circuit.call(config.null_response, config.circuit_open_error)
+        else
+          begin
+            response = yield
+            server_maintenance_timeout = response.headers[config.maintenance_mode_header].to_i
+
+            if server_maintenance_timeout.positive?
+              record_failure!
+              open!(
+                {
+                  expires_in: server_maintenance_timeout,
+                  # for logging purposes. can't log expires_in because it might be overridden if greater than max
+                  server_maintenance_timeout: server_maintenance_timeout
+                }
+              )
+
+              return config.on_open_circuit.call(response, config.circuit_open_error)
+            end
+
+            record_success!
+            response
+          rescue Faraday::Error => e
+            record_failure! if e.response_status >= 500 || config.error_codes_watchlist.include?(e.response_status)
+
+            raise e
+          end
+        end
+      end
+
+      def open?
+        storage.key?(open_storage_key)
+      end
+
+      def half_open?
+        storage.key?(half_open_storage_key)
+      end
+
+      def failure_count
+        storage.get(stat_storage_key(:failure)).to_i
+      end
+
+      def success_count
+        storage.get(stat_storage_key(:success)).to_i
+      end
+
+      def tripped_count
+        storage.get(stat_storage_key(:tripped)).to_i
+      end
+
+      def failure_rate
```

Comment:
asking for understanding - why doesn't tripped_count contribute towards lowering failure rate? If request is "tripped" (which I gather means the connection is open), but the caller keeps tripping, how does the fuse close/half-close? was thinking after tripping enough times, the connection might move to half-open

Reply:
ah fuse goes to half open after the sleep window

Reply:
The fuse closes after any successful request is made. So after the open circuit timeout expires, any successful request immediately closes the fuse. If we factor in the trip count, the calculation will be wrong because we would essentially be treating tripped events as successes (lowering failure rate), and hence it will take much longer than it should for the fuse to break open again

Vice versa: if we treat tripped calls as failures (increasing failure rate), the fuse will break much faster than it should